numpy/core/src/umath/simd.inc.src

```diff
@@ -85,10 +119,12 @@ abs_ptrdiff(char *a, char *b)
  * We instead rely on i32gather/scatter_ps instructions which use a 32-bit index
  * element. The index needs to be < INT_MAX to avoid overflow. MAX_STEP_SIZE ensures this.
  */
-#define IS_BINARY_SMALL_STEPS \
+#define IS_BINARY_SMALL_STEPS_AND_NOMEMOVERLAP \
     ((abs(steps[0]) < MAX_STEP_SIZE)  && \
      (abs(steps[1]) < MAX_STEP_SIZE)  && \
-     (abs(steps[2]) < MAX_STEP_SIZE))
+     (abs(steps[2]) < MAX_STEP_SIZE)  && \
+     (nomemoverlap(args[0], steps[0]*dimensions[0], args[2], steps[2]*dimensions[0])) && \
+     (nomemoverlap(args[1], steps[1]*dimensions[0], args[2], steps[2]*dimensions[0])))
```

Comment:
This currently ensures that there is no memory overlap across the entire arrays. For large steps and large dimensions, `steps[0]*dimensions[0]` can overflow. If this check an overkill? I don't fully understand under what circumstances does input and output arrays overlap in memory.

Reply:
For `steps[0]*dimensions[0] `to overflow, `dimensions[0] ` has to be 4398046511104 (since max step size = 2097152 and `npy_intp `has a max value of 2**63 - 1). The memory required to hold such a large array with that step size is 8192 Peta Bytes of data, which I assume is unrealistic.

Reply:
code formatting: `steps[0]*dimensions[0]` -> `steps[0] * dimensions[0]`

Reply:
done.